### ISSUE TITLE
Calling function 'SignalRModule', function calls are not supported.

### ISSUE THREAD

**Issue by cinkpen** [HNeukermans/ng2-signalr#14]
I'm using the Angular-cli  to make a very basic "Hello world" app with SignalR using your module, and I keep getting this:

```
ERROR in Error encountered resolving symbol values statically. Calling function 'SignalRModule', function calls are not supported. Consider replacing the function
or lambda with a reference to an exported function, resolving symbol AppModule in C:/code/Ang2SR/src/app/app.module.ts, resolving symbol AppModule in C:/code/Ang2S
R/src/app/app.module.ts
webpack: Failed to compile.
```

My app.module:
```

import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { HttpModule } from '@angular/http';
import { AppComponent } from './app.component';
import { EchoComponent } from './echo/echo.component';
import { signalrConfiguration } from './app.config';
import { SignalRModule, SignalRConnection, ConnectionStatus, BroadcastEventListener } from 'ng2-signalr';

@NgModule({
  declarations: [
    AppComponent,
    EchoComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule,
    SignalRModule.configure(signalrConfiguration)
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }

```

app.config.ts:
```

import { SignalRConfiguration } from 'ng2-signalr';

const signalrConfig = new SignalRConfiguration();
signalrConfig.hubName = 'echoHub';
signalrConfig.qs = { user: 'Hannes' };
signalrConfig.url = 'http://localhost:44369/signalr/hubs';


export const signalrConfiguration = signalrConfig;
```

Has this got something to do with publishing metadata/AoT?
Any ideas?

**Comment by HNeukermans** [comment]
Hi, cinkpen,
thanks for posting this error. I think this error should be fixed in v1.0.3.
Can you verify?

**Comment by cinkpen** [comment]
I'm on v1.0.5 now and the update hasn't changed the error. I was also having another problem at the same that I forgot to include - when attempting a connect: 
![image](https://cloud.githubusercontent.com/assets/8832338/22734042/2e051536-edec-11e6-8267-14b69002cccc.png)

Might be related. 

Also in your ReadMe in the "How to listen for server side events" section you reference "this.connection" but that isn't mentioned anywhere else. Where/how should I define connection? At the moment I'm trying this:

```
  setup() {
    console.log("Calling connect");
    this._signalR.connect().then(value => {
      console.log("Connect done");
      console.log(value);
      console.log("Connect done - value above.");
    })
    .catch( reason =>{
      console.log(reason);
    })
  }
```

Which when added, started the original error.

**Comment by cinkpen** [comment]
Would it be at all possible for you to make the very simplest of "Hello world" application + repo that sets up and uses a SignalR with Angular2 in its most basic form? It would provide the perfect template to work from.

**Comment by zareone** [comment]
Hi!

First of all, thank you very much for putting this lib out.

I'm on v1.0.5 and also having the issue. I've tried with Angular 2.3.1 and 2.4.6 

The erorr I'm having is: 
```
Error encountered resolving symbol values statically. Calling function 'SignalRModule', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, resolving symbol AppModule in /home/user/web-app/src/app/app.module.ts, resolving symbol AppModule in /home/user/web-app/src/app/app.module.ts
errors @ client?93b6:101
sock.onmessage @ socket.js:37
EventTarget.dispatchEvent @ eventtarget.js:51
(anonymous) @ main.js:274
SockJS._transportMessage @ main.js:272
EventEmitter.emit @ emitter.js:50
WebSocketTransport.ws.onmessage @ websocket.js:35
wrapFn @ zone.js:1039
ZoneDelegate.invokeTask @ zone.js:275
Zone.runTask @ zone.js:151
ZoneTask.invoke @ zone.js:345
```

I've seen these two commits regarding AOT issues

https://github.com/HNeukermans/ng2-signalr/commit/75b483c81906b6b20cc4144b1ba718a7cdfae552
https://github.com/HNeukermans/ng2-signalr/commit/e496bcfb0cbfba290f32ceab17a77d207c690e25

Could it be that the second one isn't included in the v1.0.5? (I don't know how to see what commit the v1.0.5 release is pointing to).

I've tried the solution pdunlop proposes in the issue #9 and it works (creating a wrapper module). 

Regards,

**Comment by HNeukermans** [comment]
Hi zareone,

thx for liking the library. I've looked into the AOT issue... but I need more time to investigate as I am new to AOT and have difficulty in understanding the errors that come with it. 
Good news there is a workaround. #9.
I' m currently giving priority to working on my live demo and code documentation.
If someone knows how to fix the AOT issue, I'll be glad to merge it in...,

**Comment by HNeukermans** [comment]
@cinkpen 
both [ng2-signalr.demo.webui.systemjs](https://github.com/HNeukermans/ng2-signalr.demo.webui.systemjs) and   [ng2-signalr.demo.webui](https://github.com/HNeukermans/ng2-signalr.demo.webui) can help you out. 
I'm working on the documentation. It is work in progress  http://ng2-signalr-webui.azurewebsites.net/

**Comment by HNeukermans** [comment]
found this commit inside angular2 apollo github project that solved there AOT issue
https://github.com/apollographql/apollo-angular/pull/195/commits/d051fee4a57cab16342a26581b5bca867e7b37dc

**Comment by DaveMonag** [comment]
Any luck integrating this fix yet?

**Comment by HNeukermans** [comment]
Hi DaveMonag,

not yet. Focussing on test coverage for the moment...

Apparently, this feature is highly requested...
Will need to look into...

**Comment by HNeukermans** [comment]
@NgModule({
    declarations: [
        TranslatePipe,
        TranslateDirective
    ],
    exports: [
        TranslatePipe,
        TranslateDirective
    ]
})
export class TranslateModule {
    /**
     * Use this method in your root module to provide the TranslateService
     * @param {TranslateModuleConfig} config
     * @returns {ModuleWithProviders}
     */
    static forRoot(config: TranslateModuleConfig = {}): ModuleWithProviders {
        return {
            ngModule: TranslateModule,
            providers: [
                config.loader || {provide: TranslateLoader, useClass: TranslateFakeLoader},
                config.parser || {provide: TranslateParser, useClass: TranslateDefaultParser},
                config.missingTranslationHandler || {provide: MissingTranslationHandler, useClass: FakeMissingTranslationHandler},
                TranslateStore,
                {provide: USE_STORE, useValue: config.isolate},
                TranslateService
            ]
        };
    }

    /**
     * Use this method in your other (non root) modules to import the directive/pipe
     * @param {TranslateModuleConfig} config
     * @returns {ModuleWithProviders}
     */
    static forChild(config: TranslateModuleConfig = {}): ModuleWithProviders {
        return {
            ngModule: TranslateModule,
            providers: [
                config.loader || {provide: TranslateLoader, useClass: TranslateFakeLoader},
                config.parser || {provide: TranslateParser, useClass: TranslateDefaultParser},
                config.missingTranslationHandler || {provide: MissingTranslationHandler, useClass: FakeMissingTranslationHandler},
                {provide: USE_STORE, useValue: config.isolate},
                TranslateService
            ]
        };
    }
}

**Comment by HNeukermans** [comment]
I'm going to wait on solving this issue. Apparantly there is a bug in the AOT  compiler, that ignores the main property inside the package.json. Therefore, it is needed to compile your library to the root folder, for the sake of AOT compat. Currently my library compiles to /lib folder. I prefer to wait for the fix to come....

**Comment by DaveMonag** [comment]
Ok, as none of the temporary fixes seemed to work for me, in order to build I've done the following in order to get things working. I absolutely do not suggest this fix, but if you just desperately need it to work, this does. **Essentially I've entirely bypassed using the "SignalRModule"** from ng2-signalr entirely, and just gone around it and that fixes the problem.

First off I **cloned the repo from here** rather than used "npm install ng2-signalr" as I needed the typescript. Then I moved that module into my app work space for ease of use. 

in your app.module.ts, remove 
`SignalRModule.configure(configuration)`
from your imports, and in providers, add
```
  providers: [
    SignalR,
    SignalRConfiguration
  ],
```
and at the top import them from the new module directory. E.g.:

```
import { SignalRConfiguration } from './modules/ng2-signalr';
import { SignalR } from './modules/ng2-signalr';
```
At this point you might want to move everything out of "lib" so that everything is just under ng2-signalr as opposed to "ng2-signalr/src".
Then in the new module, go into the signalr.ts and replace the constructor to this and add these three methods:

```
    public constructor(configuration: SignalRConfiguration, zone: NgZone) {

        this._configuration = configuration;
        this._zone = zone;
        this._jHubConnectionFn = this.getJConnectionFn();
    }

    private getJConnectionFn(): any {
        let jQuery = this.getJquery();
        let hubConnectionFn = (<any>window).jQuery.hubConnection;
        if (hubConnectionFn == null) throw new Error('Signalr failed to initialize. Script \'jquery.signalR.js\' is missing. Please make sure to include \'jquery.signalR.js\' script.');
        return hubConnectionFn;
    }

    private getJquery(): any {
        let jQuery = (<any>window).jQuery;
        if (jQuery == null) throw new Error('Signalr failed to initialize. Script \'jquery.js\' is missing. Please make sure to include jquery script.');
        return jQuery;
    }

    public configure(configuration: SignalRConfiguration) {
        this._configuration = configuration;

    }
```

Now I've put my config in a file called "app.config.ts", which you can still use in this scenario:
```
import { SignalRConfiguration } from './modules/ng2-signalr';

export const signalrConfiguration = new SignalRConfiguration();

signalrConfiguration.hubName = 'yourHub';
signalrConfiguration.qs = { user: 'user' };
signalrConfiguration.url = 'http://whateveryourbackendis.azurewebsites.net';
```

I've been handling all the SignalR interactions in my app inside a service. In this service, in the constructor, I've added a call to the "configure" method we just added in _signalR:
```
  constructor(private _signalR: SignalR) {
    this._signalR.configure(signalrConfiguration);
    this._OnReceiveMessage = new BroadcastEventListener<Alarm>("message");
    this.setup();
  }
```

So obviously the problem for us lies with the SignalRModule, as just bypassing it all together and leaving everything else fine allows everything to work just fine. This obviously isn't a permanent fix, but perhaps you can make something of this @HNeukermans .

**Comment by HNeukermans** [comment]
Excellent stuff @DaveMonag.
You seem to be an expert :-). Much appreciate your help on this.

**Comment by eshvatskyi** [comment]
First of all ))) thanks for lib. nice one.
`export class SignalR`
`{`
`    ...`
`    private _jHubConnectionFn: Function;`
``
`    public constructor(configuration: SignalRConfiguration, zone: NgZone, jHubConnectionFn: Function)`

it's pretty enought to make it compatible with AOT

**Comment by HNeukermans** [comment]
Hi @eshvatskyi,

I don't understand your comment 100% correctly. Could you be more specific.

**Comment by DaveMonag** [comment]
In the constructor in the SignalR.ts file, you have written a lambda function. His code removes that lambda function. The original error complained of a lambda function in a constructor as the source of the problem. I tried changing that before and it didn't fix anything for me, perhaps I should retry.

**Comment by eshvatskyi** [comment]
My fix works on latest ng and cli. + i18n also works as well

**Comment by DaveMonag** [comment]
When I try this fix, the file that has the error just seems to change from the module to the configuration file:

```
ERROR in Error encountered resolving symbol values statically. Calling function 'SignalRConfiguration', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, resolving symbol AppModule in C:/Users/code/deleteMePlease/src/app/app.module.ts, resolving symbol AppModule in C:/Users/code/deleteMePlease/src/app/app.module.ts, resolving symbol AppModule in C:/Users/code/deleteMePlease/src/app/app.module.ts
webpack: Failed to compile.
```

I figured it was the same problem before and dismissed this to continue looking. Any ideas?

**Comment by HNeukermans** [comment]
I'm going to retry aot, now that angular 4 is out...

**Comment by alcfeoh** [comment]
Perhaps all you need to do is rename the `configure` function to `forRoot ` as suggested by the Angular convention here: [https://angular.io/docs/ts/latest/cookbook/ngmodule-faq.html#!#q-for-root](https://angular.io/docs/ts/latest/cookbook/ngmodule-faq.html#!#q-for-root)

**Comment by dylabar** [comment]
@HNeukermans,

I've been writing AOT Angular libraries that integrate with the latest version of Angular CLI, and was in the process of writing one that relies on SignalR. Thus I stumbled on your project, but am unable to AOT build my library that is utilizing yours due to the issue described in this post.

However, I've run into very similar problems. While I'm still searching for the ideal solution, I have found something that works, but can be improved upon. I've made some minor modifications to your SignalRModule based on how I've set up my libraries. Keep in mind I have not tested these changes, I'm simply following how I've implemented it elsewhere.

```
export class SignalRModule {
    public static forRoot(getSignalRConfiguration: Function): ModuleWithProviders {
        return {
            ngModule: SignalRModule,
            providers: [
                {
                    provide: SIGNALR_CONFIGURATION,
                    useFactory: getSignalRConfiguration
                },
                {
                    provide: SignalR,
                    useFactory: (createSignalr),
                    deps: [SIGNALR_CONFIGURATION, NgZone]
                }
            ],
        };
    }
}
```

A few things to notice here:

1. I am using the conventional "forRoot" function instead of "configure".
2. A function is being passed in as opposed to a class or interface.
3. The function is provided to the rest of the library via a provider and the "useFactory" convention.

The downside, and what I am currently working on improving, is the fact that the parameter is typed broadly to function. If I could type it to a function of a specific return type (for instance, a function that has to return a SignalRConfiguration), this would be an improvement.

Here is a code example on how a user would then provide the configuration in their app:

```
export function createSignalRConfiguration(): SignalRConfiguration {
  return {
    url: '',
    hubName: '',
    logging: false
  } as SignalRConfiguration ;
}

@NgModule({
  imports: [
    SignalRModule.forRoot(createSignalRConfiguration)
  ]
})
```

I'm hoping these examples will help point you in the right direction of making this application AOT compatible. We would love to use your library :)

**Comment by dylabar** [comment]
I've begun an attempt to implement my changes and test them. While the changes described in my previous post got past the initial error, I am receiving a new AOT issue with the signalr.ts file. I will continue to work on this, and submit a pull request if I can get it working. However, I have limited knowledge on how this library operates internally, so I may hit a road block at some point.

**Comment by DaveMonag** [comment]
@dylabar Could you post your getSignalRConfiguration function please?

**Comment by dylabar** [comment]
@DaveMonag The "getSignalRConfiguration" function is the name of the forRoot parameter. In the second code snippet, I use the function "createSignalRConfiguration" and pass it in as this parameter. Therefore I believe the function you're looking for is actually the createSignalRConfiguration function.

**Comment by DaveMonag** [comment]
Ah, sorry. Now I just look silly for not seeing that, my bad. 

YES! That's fixed it for me! Not getting errors anywhere anymore! 
I've just done @dylabar 's fix on top of @eshvatskyi 's and it worked. I haven't yet tested if eshvatskyi's fixes are actually necessary though. Either way, good job guys.

**Comment by dylabar** [comment]
Thank you for the confirmation @DaveMonag :). You were right, @eshvatskyi 's fix resolves the secondary error I received in a previous post I made. I have implemented both code changes in a fork, and submitted pull request #28.

**Comment by DaveMonag** [comment]
Just tried

**Comment by itsurge** [comment]
I have this issue !! But I read this topic  and don't get solution Please explain me the solution step by step,
thanks

**Comment by HNeukermans** [comment]
The long awaited fix in on the way.  I already merged in the request. Still need to publish it to npm and update the readme of the breaking change...  Thx to @dylabar and @DaveMonag.

...some patience

**Comment by HNeukermans** [comment]
Hi,
@dylabar and @DaveMonag, could you link me to a working solution, bc I can't get it working...

**Comment by dylabar** [comment]
@DaveMonag may have a good example for you from the sounds of it, any example I have would be in a private work project. Does the package still build properly? If the issues you are having are in the client, are you receiving any error messages? Here is a snippet of what importing the service looks like in my app:

```
export function createSignalRConfiguration(): SignalRConfiguration {
  return {
    url: '',
    hubName: '',
    logging: false
  } as SignalRConfiguration ;
}

@NgModule({
  imports: [
    SignalRModule.forRoot(createSignalRConfiguration)
  ]
})
```

**Comment by LatisVlad** [comment]
Hi, any updates on this? Thanks!

**Comment by DaveMonag** [comment]
My service looks like this, its barebones but it works:
```
import { Injectable } from '@angular/core';
import { SignalRConnectionBase, SignalR, BroadcastEventListener } from 'ng2-signalr';
import { Subject } from 'rxjs/Subject'
import { Message } from './Message'


@Injectable()
export class Signalrconnector {
  // -- SignalR -- 
  private _Connection: SignalRConnectionBase;
  private _OnReceiveMessage: BroadcastEventListener<Message>
  public messageSubject: Subject<any> = new Subject<any>();

  constructor(private _signalR: SignalR) {
    console.log(this._signalR);
    this._OnReceiveMessage = new BroadcastEventListener<Message>("NotifyUser");
    this.setup();
  }

  setup(): Promise<any> {
    var p = new Promise((resolve, reject) => {
      // Initialise connection, and execute function once connected

      this._signalR.connect().then(connection => {
        this._Connection = connection;
        // Start the connection
        connection.start();
        // Subscribe to the event listener and callback when new message received
        this._OnReceiveMessage
          .subscribe(p => {
            console.log(p);
            this.messageSubject.next(p)
          });
        // Start listening to listener
        this._Connection.listen(this._OnReceiveMessage);
        // Resolve empty to resolve P - no need to pass any parameters
        resolve();
      })
    });
    // return to tell anything awaiting promise that setup has finished
    return p;
  }
}
```

and module looks like this:

```
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { HttpModule } from '@angular/http';

import { AppComponent } from './app.component';
import { SignalRModule } from 'ng2-signalr';
import { SignalRConfiguration } from 'ng2-signalr';

export function createSignalRConfiguration(): SignalRConfiguration {
  return {
    url: 'http://notmyactualsignarhublocation.azurewebsites.net',
    hubName: 'messageHub',
    logging: true
  } as SignalRConfiguration ;
}

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule,
    SignalRModule.forRoot(createSignalRConfiguration)
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```
This is from a mock project I've made so that I can share this kind of thing, as I'm working on a company project at the moment hence cant share the code directly
I've tested with the version in GitHub with this project, it works for me. But the current npm version currently doesn't contain the fixes, so breaks. What problem are you having @HNeukermans ?
Also sorry for late reply, was on a short holiday.

**Comment by DaveMonag** [comment]
In order to do this with the GitHub clone, I'm taking the "src" file, renaming that "ng2-signalr", placing it in node modules, removing the testing folder, and then removing all the spec.ts files too. Also last thing to remember is removing the "export * from './testing" line from the index.ts file within the folder. Then with my code above, it works.

**Comment by HNeukermans** [comment]
After the lastest code changes + configuration according to DaveMonag, still seem to get the makeDecorator ... exception. This expection is really useless. The exception refers to something in angular/core blabbla bla.. .AOT error seem to drive me crazy. Probably related to not publishing my lib in the right way. Need to also publish the metadata files...

**Comment by MarkDeVerno** [comment]
Just checking in to see if there are any updates coming for this.  This issue is pretty breaking as projects can't be built or deployed since compilation fails.

**Comment by HNeukermans** [comment]
issue fixed in v2.0.0
